### ISSUE TITLE
Slice is not subscriptable at runtime

### ISSUE THREAD

**Issue by joecox** [astral-sh/ty#1595]
### Summary

ty should report [`not-subscriptable`](https://docs.astral.sh/ty/reference/rules/#non-subscriptable) when annotating slice types below 3.14:

```sh
$ cat >slice.py <<'# EOF'
s: slice[int]
# EOF

$ uv run --python 3.13 slice.py
Using CPython 3.13.7 interpreter at: /opt/homebrew/opt/python@3.13/bin/python3.13
Removed virtual environment at: .venv
Creating virtual environment at: .venv
Traceback (most recent call last):
  File "/Users/joe/slice.py", line 1, in <module>
    s: slice[int]
       ~~~~~^^^^^
TypeError: type 'slice' is not subscriptable

$ uv run --python 3.14 slice.py
Using CPython 3.14.0
Removed virtual environment at: .venv
Creating virtual environment at: .venv

$ uvx --python 3.13 --with "ty==0.0.1-alpha.27" ty check slice.py
All checks passed!
```

I assume ty doesn't complain because typeshed has slice as generic? I don't even know if it's documented anywhere that subscripting slice began working on 3.14.

### Version

ty 0.0.1-alpha.27 (26d7b6864 2025-11-18)

**Comment by carljm** [comment]
Thanks for the report!

In fact, `slice` didn't become subscriptable at runtime on Python 3.14 either. What happened was that type annotations became deferred by default (PEP 649/749), so `s: slice[int]` doesn't execute the code `slice[int]` at runtime.

```
Python 3.14.0rc2 (main, Aug 28 2025, 17:02:21) [Clang 20.1.4 ] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> slice[int]
Traceback (most recent call last):
  File "<python-input-0>", line 1, in <module>
    slice[int]
    ~~~~~^^^^^
TypeError: type 'slice' is not subscriptable
```

Ideally, type checkers should not assume that just because `slice` is generic in typeshed, that also means it is subscriptable at runtime. Runtime subscriptability should be indicated separately by defining `__class_getitem__`. But in typeshed, `Generic` itself defines `__class_getitem__`, so it may be difficult for type checkers to make this distinction.

None of the other type checkers (mypy, pyright, pyrefly) report this either.

So I think it will not be a priority for us in the near term, but I'll keep the issue open as something we may want to look at in future.

**Comment by joecox** [comment]
That makes sense, thanks for the info!

**Comment by AlexWaygood** [comment]
> Ideally, type checkers should not assume that just because `slice` is generic in typeshed, that also means it is subscriptable at runtime. Runtime subscriptability should be indicated separately by defining `__class_getitem__`.

Yes, the reason why we generally include `__class_getitem__` methods in typeshed when they exist at runtime (even for classes that inherit from `Generic` in typeshed) is so that type checkers can have the option of distinguishing between classes that are generic at runtime _and_ at type-check time, and those which are only generic at type-check time.

> But in typeshed, `Generic` itself defines `__class_getitem__`, so it may be difficult for type checkers to make this distinction.

What do you mean here? I don't think typeshed says anything about whether `Generic` defines `__class_getitem__`. `Generic` isn't even a class in typeshed.

> None of the other type checkers (mypy, pyright, pyrefly) report this either.
> 
> So I think it will not be a priority for us in the near term, but I'll keep the issue open as something we may want to look at in future.

Yeah, I agree that this shouldn't be a priority

**Comment by carljm** [comment]
> What do you mean here? I don't think typeshed says anything about whether `Generic` defines `__class_getitem__`. `Generic` isn't even a class in typeshed.

I was looking at https://github.com/python/typeshed/blob/main/stdlib/typing.pyi#L445-L457. Admittedly I didn't think very hard about why it does that funny dance with `class _Generic` and then `Generic: type[_Generic]` -- but I think that if taken literally (and if a type checker allowed subclassing a `type[...]` type), that would still have the effect of making a type checker think that a class inheriting `Generic` had `__class_getitem__` (at least on Python 3.10+).

That said, I don't think that's actually the issue in ty; we special-case inheritance from Generic anyway. I agree that if this is true:

> we generally include `__class_getitem__` methods in typeshed when they exist at runtime (even for classes that inherit from `Generic` in typeshed)

then in principle type checkers have the information they need in order to handle this correctly (even though apparently they don't!).

The issues in ty here are really two-fold:

1) We don't type check annotations as runtime value expressions, on any Python version, even though ideally we should for Python <3.13, without `from __future__ import annotations`. This might be hard to fix with our current approach to value vs type expressions, but I think we should change our approach there in a way that would make it easier in the future.
2) Even in value expressions, we special-case explicit subscription of generic classes and always allow it, regardless of whether the class has `__class_getitem__` or not. This would be pretty easy to fix, I think.

**Comment by AlexWaygood** [comment]
> I was looking at https://github.com/python/typeshed/blob/main/stdlib/typing.pyi#L445-L457. Admittedly I didn't think very hard about why it does that funny dance with `class _Generic` and then `Generic: type[_Generic]` -- but I think that if taken literally (and if a type checker allowed subclassing a `type[...]` type), that would still have the effect of making a type checker think that a class inheriting `Generic` had `__class_getitem__` (at least on Python 3.10+).

Oh! I'm sorry, I forgot the definition changed in https://github.com/python/typeshed/pull/14583; I had the previous definition still in my head (`Generic` used to just be an instance of `_SpecialForm` in typeshed). The funny dance is because pyright broke in the first version of that PR, when we tried to make `Generic` an actual class.

**Comment by adamjstewart** [comment]
Is there an open issue where I can track support for Python 3.15 generic slice support or should I open a new one? I don't see it listed in #1889.

**Comment by AlexWaygood** [comment]
@adamjstewart it doesn't need any special support from us, really; we should already support it. This issue is about the fact that we already consider slice as being generic on all Python versions, because that's what typeshed tells us. That's not quite accurate on Python <=3.14 (which is why this issue is open), but it is accurate on a Python 3.15.
